OSGi Service as Global EPackageRegistry

juergen-albert opened the pull request:
The default EPackage Registry can be overwritten in mutlple ways (System property or ExtensionPoint). With this, if non of the other mechanisms apply, there now can be a service that acts as the static Registry.

The Mechanism makes sure that we will always have a functional and correctly filled static registry. All write commands, will be held in a backup registry, which will be reapplied if the service comes or goes. If multiple services are available, the one with the highest service ranking will win.

@merks I have a few tests for this, but I couldn't find tests in the Project, that have a running OSGi. If you want to have a look at the tests, they are [here](https://github.com/eclipse-fennec/emf.osgi/blob/emf_osgi_delegate_global_registry/org.eclipse.emf.ecore.test/src/org/eclipse/emf/ecore/test/GlobalEPackageServiceRegistryTest.java).

merks commented:
All the tests run as OSGi applications.   But I guess you want to launch in some special way and it looks like you have some more special fancy test infrastructure/libraries that you use...

merks commented:
I've not looked in great detail at the following PR which seems somewhat related at a glance:

https://github.com/eclipse-emf/org.eclipse.emf/pull/52

juergen-albert commented:
> I've not looked in great detail at the following PR which seems somewhat related at a glance:

Related but something different. If I understand Hannes PR correctly, it collects .ecore files that are marked in the Manifest of a Bundle and registers the packages, similar to what the generated code does. I however want to have a possiblity to replace the static package registry completly, without the use of the system property or the Extensionpoint.

juergen-albert commented:
> All the tests run as OSGi applications. But I guess you want to launch in some special way and it looks like you have some more special fancy test infrastructure/libraries that you use...

Am I blind? Where can I find them? My tests are not really that special and should be adaptable to your tests with minimal effort.

merks commented:
There's this:

https://github.com/eclipse-emf/org.eclipse.emf/tree/master/tests/org.eclipse.emf.test.core

The build launches tests like this:
```
[INFO] --- tycho-surefire:4.0.10:test (default-test) @ org.eclipse.emf.test.core ---
[INFO] Executing test runtime with timeout (seconds): 0, logs, if any, will be placed at: /home/jenkins/agent/workspace/build_master/tests/org.eclipse.emf.test.core/target/work/data/.metadata/.log
[INFO] Command line:
	[/opt/tools/java/openjdk/jdk-21/latest/bin/java, -Dosgi.noShutdown=false, -Dosgi.os=linux, -Dosgi.ws=gtk, -Dosgi.arch=x86_64, -Xmx1024m, -ea, -Dorg.eclipse.emf.common.util.ReferenceClearingQueue=true, -Dosgi.clean=true, -jar, /home/jenkins/.m2/repository/p2/osgi/bundle/org.eclipse.equinox.launcher/1.6.900.v20240613-2009/org.eclipse.equinox.launcher-1.6.900.v20240613-2009.jar, -data, /home/jenkins/agent/workspace/build_master/tests/org.eclipse.emf.test.core/target/work/data, -install, /home/jenkins/agent/workspace/build_master/tests/org.eclipse.emf.test.core/target/work, -configuration, /home/jenkins/agent/workspace/build_master/tests/org.eclipse.emf.test.core/target/work/configuration, -application, org.eclipse.tycho.surefire.osgibooter.headlesstest, -testproperties, /home/jenkins/agent/workspace/build_master/tests/org.eclipse.emf.test.core/target/surefire.properties]
```
The IDE should look like this:

![image](https://github.com/user-attachments/assets/f78e2945-e062-4c67-a3c4-1be39089a0de)

juergen-albert commented:
Thanks heaps. Before I put in more work, by migrating my Tests: What is your general feeling towords this PR? Would this be something you would accept, given the tests and details work out?

merks commented:
I really need some time to review this before I can have a well-informed opinion and I am horribly swamped these days and I'm out of office next week...

I will definitely get fussy at some point about a space between `if` and `(` and about matching the brace style used by the rest of the code but that's simple superficial stuff...

juergen-albert commented:
> I really need some time to review this before I can have a well-informed opinion and I am horribly swamped these days and I'm out of office next week...

No Problem and thank you.

>I will definitely get fussy at some point about a space between if and ( and about matching the brace style used by the rest of the code but that's simple superficial stuff...

I have tried to match your code style as best, as I was able. I imported the project via the Oomph setup, but there have been no codestyles attached. Do you have anything I can use or was I doing anything wrong during installation?

merks commented:
There's this:

![image](https://github.com/user-attachments/assets/2d8b76fd-8e13-4903-ad21-48c2bff29ac8)

But I don't generally format entire files, except for completely new ones.  I.e., I prefer to preserve the history...

juergen-albert commented:
@merks did you find the time to take a look at the PR in general?

juergen-albert commented on plugins/org.eclipse.emf.ecore/src/org/eclipse/emf/ecore/plugin/EcorePlugin.java:
```diff
@@ -656,10 +663,60 @@ public Implementation()
+      try {
+        ExtensionProcessor.internalProcessExtensions();
+      } catch (NoClassDefFoundError e) {
+        // If EMF runs in a non Equinox OSGi Framework, it occasionally runs in some NoClassDefFoundError Exceptions, that can safely be ignored.  
```
First of: EMF works perfectly fine in Felix! So there is in general nothing to worry about on your end here. 

However :-), we are using QVT quite a lot and we are in some cases bound to Felix. QVT however has never gotten to the point where you are, so it can run without the usual `org.eclipse.core.runtime` dependency. Even with the supplement is does not work. Thus, we have to jump through a lot of hoops and provide a really scetchy Bundle, that sells QVT the notion, that it has everything it needs. This unfortunately has the consequence, that the `EMFPlugin.IS_ECLIPSE_RUNNING` suddenly also thinks it is true. This leads in the end to it expecting the ExtensionRegistry to be around, which we don't have and never will have.

We are currently working on QVT, to become independent of Equinox, but we don't know when and how this will be finished. Thus this is more of a precaution, that the Activator runs though, even if we run in scetchy mode.

juergen-albert commented on plugins/org.eclipse.emf.ecore/src/org/eclipse/emf/ecore/plugin/EcorePlugin.java:
```diff
@@ -668,7 +725,12 @@ public static final class Activator extends EMFPlugin.OSGiDelegatingBundleActiva
+        {
+          return new Implementation();
+        } catch (Throwable t) {
+          return new PlainOSGiBundleActivator();
```
Well, I only used what you already provided. When EMF runs in Felix, it will always run into the catch block, as a `NoClassDefFoundError` will be thrown when the `createBundle` method is called, as in Felix you will not find the `EclipsePlugin` Implementation inherits from. 

Thinking about it, it might be a better solution in general to use the `EMFPlugin.IS_ECLIPSE_RUNNING` as an indicator on what todo, instead of accepting the `NoClassDefFoundError`.

juergen-albert replied:
I'm not sure what you mean? `createBundleHelper` catches the Throwable, but I'm not sure what I can do about it, as this is the mechanism that was in place before. We can at least log an error in the `OSGiDelegatingBundleActivator` in such a case.

juergen-albert commented on plugins/org.eclipse.emf.ecore/src/org/eclipse/emf/ecore/plugin/EcorePlugin.java:
```diff
@@ -656,10 +663,60 @@ public Implementation()
+     * In Case we are not running in Eclipse, this BundleActivator will be used
+     * @since 2.40
+     */
+    public static class PlainOSGiBundleActivator implements BundleActivator {
```
The `Implementation` you provide, extends `EclipsePlugin` and will fail outside Eclipse. Thus I need a "Plain" BundleActivator, that will always work. Thus, this is a separate Class, which is used by the "Eclipse" Activator or standalone if we are not running in Eclipse.

juergen-albert commented:
Maybe some general Context. The way I implemented this, nothing should change for anybody. In normal JavaSE nothing changes. In any OSGi Environment, there will always be a default Registry, that can be overwritten via the System property or ExtensionPoint. Thus the existing mechanisms will always take precedence over the new mechanism.

We however don't have any extension points and never will have them. Thus there is no way for us to provide our own registry.

In my opinion the static registry should be used by generated code or ecore deployed with a Bundles.  We however have a System, where besied the generated Code, EPackages can dynamically come and go and must be available without a ResourceSet around. Thus I need a way to exchange the Registry. 

I understand worries about the reentrent behavior. This is the reason, I introduced the lock and the only moment it blocks, is during the arrival of a new PackageRegistry Service. If one goes down the rabbit hole, to provide your own Registry, you are already in dangerous territory anyway and can screw up the whole system.

merks commented:
I think it always replaces the registry with your new registry:

![image](https://github.com/user-attachments/assets/b347004d-669c-4d1f-9041-3b2214ed9b01)

What if we provided a way to specify via a system property both the bundle name and the class name, `<bsn>:<class-name>`,  so that it can be created reflectively from the correct bundle.  

I'd prefer to introduce the least amount of change with the least risk of regression with the least amount of new code to maintain.

juergen-albert commented:
This would not necessarily work, as we can't guarantee that the bundle, the property would name is already installed and resolved at the time this code is called.

juergen-albert commented:
We could use a system property as a feature toggle and have the part deactivated by default.

merks commented:
> We could use a system property as a feature toggle and have the part deactivated by default.

I'm really uncomfortable with not understanding exactly was is present in the runtime and what is not present, what's failing, why it's failing, and so on.  Certainly clearly documented system property guards that have no impact by default would make me feel less like I'm risking regressions I don't even understand.

Does the class org.eclipse.emf.ecore.plugin.OSGiDelegatEPackageRegistry need to be public API or could it be completely hidden, e.g., as a private static class in EcorePlugin? I would feel better to make as little new API commitment as possible...

juergen-albert commented:
> I'm really uncomfortable with not understanding exactly was is present in the runtime and what is not present, what's failing, why it's failing, and so on. Certainly clearly documented system property guards that have no impact by default would make me feel less like I'm risking regressions I don't even understand.

I will make the change.

juergen-albert commented:
> Does the class org.eclipse.emf.ecore.plugin.OSGiDelegatEPackageRegistry need to be public API or could it be completely hidden, e.g., as a private static class in EcorePlugin? I would feel better to make as little new API commitment as possible...

Not sure why I made it public. I can make it an internal inner class.

merks commented:
Yes, that's unrelated and something that I fixed just a while back

https://github.com/eclipse-emf/org.eclipse.emf/commit/46bae4e27831fb734bb6ebd92f369034b9ecb3f0

and the master build does work:

https://ci.eclipse.org/emf/job/build/job/master/478/

Anyway I must run now...

I'll look at the PR tomorrow and run a local Tycho build...

juergen-albert commented:
Why the hell, don't I get any notifications when something happens to this PR...

juergen-albert commented on tests/org.eclipse.emf.test.core/Test EMF Core.launch:
```diff
@@ -36,7 +36,7 @@
     <stringAttribute key="org.eclipse.jdt.launching.PROJECT_ATTR" value="org.eclipse.emf.test.core"/>
     <stringAttribute key="org.eclipse.jdt.launching.SOURCE_PATH_PROVIDER" value="org.eclipse.pde.ui.workbenchClasspathProvider"/>
-    <stringAttribute key="org.eclipse.jdt.launching.VM_ARGUMENTS" value="-Dorg.eclipse.emf.test.core.ecore.EcoreValidationTest.validateAllRegisteredModels=false  -ea"/>
+    <stringAttribute key="org.eclipse.jdt.launching.VM_ARGUMENTS" value="-Dorg.eclipse.emf.test.core.ecore.EcoreValidationTest.validateAllRegisteredModels=false -Dorg.eclipse.emf.ecore.EPackage.Registry.OSGI_SERVICE=true -ea"/>
```
I have an issue with the testing. In order to properly make sure to test both Versions, I believe we would have to run the tests at least twice. One round with the Flag active and on with without. It would be even better to have a third test run, with e.g. a Felix instead of the Equinox with Eclipse. I hove no clue how to do this though with tycho. Do you?

merks commented:
Yes, something I get the sense that I miss notifications, e.g.,. when a PR is opened or a question is asked...

FYI, I think the runway has grown to short.  Next week I have to make a release build for the Platform RC2...

merks commented:
Running the tests multiple times is feasible/reasonable.  Testing against Felix would be an interesting improvement.

With the release cycle winding down, and with other demands on my time, I dropped the ball a bit here.  Sorry about that. ☹️

merks commented:
I think this is heading in the right direction now...

I have a general question now about what's available and what is not.

Specifically I think maybe the org.eclipse.core.runtime is present.  But I think you are trying to exclude the org.eclipse.equinox.registry bundle.  The historical support for running on Felix was designed to include that bundle.  I ask because I think the class org.eclipse.core.runtime.Plugin is available because org.eclipse.core.runtime.Platform is available from the same bundle.  As suchI think we don't need a separate PlainOSGiBundleActivator but rather need a guard that avoid calling 

![image](https://github.com/user-attachments/assets/dd51ca0c-16ed-45dd-af3d-a58a4578a879)

and fold the implementation details into the base class.

And then maybe we can use USE_OSGI_SERVICE_AS_EPACKAGE_REGISTRY as the guard for the above (it should be private and final) and don't need to change the meaning/implementation of IS_ECLIPSE_RUNNING.  Or might one want the service-based registry when Eclipse is really running?

Anyway, I think it's headed in a more comfortable direction...

merks commented on plugins/org.eclipse.emf.ecore/src/org/eclipse/emf/ecore/plugin/EcorePlugin.java:
```diff
@@ -586,13 +603,18 @@ public static Map<String, URI> getEPackageNsURIToDynamicModelLocationMap(boolean
-    /**
+    
+    
+    private PlainOSGiBundleActivator plainOSGiBundleActivator;
```
If I understand the current state correctly IS_ECLIPSE_RUNNING will be false we don't really need this case anymore.  Or?

juergen-albert replied:
We would still need an additional Activator, in case `IS_ECLIPSE_RUNNING` is false and `IS_OSGI_RUNNING` is true. In this case, we need an activator but we don't have the `EclipsePlugin`. We could copy the stuff from the `PlainOSGiBundleActivator` in here as well, but we would end up with duplicated code.

merks reviewed (changes requested):
There are a number of hole in the locking via the views that are returned.  Perhaps all those views should be of an modifiable view of the underlying registry.

Maybe you could try to address some of these issues before try to take this further.

juergen-albert commented on plugins/org.eclipse.emf.ecore/src/org/eclipse/emf/ecore/plugin/EcorePlugin.java:
```diff
@@ -657,9 +679,61 @@ public void start(BundleContext context) throws Exception
+      public void start(BundleContext context) throws Exception {
+        if(getDefaultRegistryImplementation() != null && getDefaultRegistryImplementation() instanceof OSGiDelegatEPackageRegistry) 
+        {
+          osgiEpackageRegistryTracker = new ServiceTracker<>(context, OSGiDelegatEPackageRegistry.FILTER, (ServiceTrackerCustomizer<EPackage.Registry, Object>) getDefaultRegistryImplementation());
```
Makes sense. I can change that.

juergen-albert commented on plugins/org.eclipse.emf.ecore/src/org/eclipse/emf/ecore/plugin/EcorePlugin.java:
```diff
@@ -1309,4 +1395,412 @@ private static void computeModels(Map<URI, URI> pluginMap, Map<String, URI> nsUR
+    public void putAll(Map<? extends String, ? extends Object> m) {
+      lock.readLock().lock();
+      try {
+        if (delegate == null) {
+          backup.putAll(m);
+        } else {
+          delegate.putAll(m);
+        }
```
Ohh, that seems like a classic copy and waste error. Yes it should all ways call putAll on backup. 
```suggestion
        backup.putAll(m);
        if (delegate != null) {
          delegate.putAll(m);
        }
```

juergen-albert commented on plugins/org.eclipse.emf.ecore/src/org/eclipse/emf/ecore/plugin/EcorePlugin.java:
```diff
@@ -1309,4 +1395,412 @@ private static void computeModels(Map<URI, URI> pluginMap, Map<String, URI> nsUR
+    public void clear() {
+      lock.readLock().lock();
+      try {
+        if (delegate == null) {
+          backup.clear();
+        } else {
+          delegate.clear();
+        }
```
```suggestion
        backup.clear();
        if (delegate == null) {
          delegate.clear();       
        }
```

juergen-albert commented on plugins/org.eclipse.emf.ecore/src/org/eclipse/emf/ecore/plugin/EcorePlugin.java:
```diff
@@ -1309,4 +1395,412 @@ private static void computeModels(Map<URI, URI> pluginMap, Map<String, URI> nsUR
+    public Collection<Object> values() {
+      lock.readLock().lock();
+      try {
+        if (delegate == null) {
+          return backup.values();
+        } else {
+          return delegate.values();
+        }
```
The idea of the lock here is to only block, when a new delegate arrives or a delegate goes away/is replaced. If we lock any write operations, we will end up with deadlocks, when geenrated bundles are intialized.

juergen-albert commented on plugins/org.eclipse.emf.ecore/src/org/eclipse/emf/ecore/plugin/EcorePlugin.java:
```diff
@@ -1309,4 +1395,412 @@ private static void computeModels(Map<URI, URI> pluginMap, Map<String, URI> nsUR
+    }
+
+    private void handleDelegateRegistryChange() {
+      Collections.sort(knownReferences,
+          Comparator.comparingInt(this::getServiceRank).thenComparing(Comparator.reverseOrder()));
+      Collections.reverse(knownReferences);
+      if (knownReferences.isEmpty() && delegate != null) {
```
Not sure, why I did that and it makes absolutly no sense... I should reverse the lists twice more, just to make sure :-)

juergen-albert replied:
```suggestion
      Collections.sort(knownReferences,
          Comparator.comparingInt(this::getServiceRank));
      if (knownReferences.isEmpty() && delegate != null) {
```

juergen-albert commented:
@merks I again did not receive any notifications and thus missed your comments. I tried unsubscribe and then resubscribed. Lets see if, if it works better now.

The last time we talked, you mentioned you wanted to take the PR and wrangle it, into something you like. Thus the question arises, if I should make the changes on the parts you have commented on?

merks commented:
@juergen-albert 

I've created the following PR that supersedes this PR:

https://github.com/eclipse-emf/org.eclipse.emf/pull/66
